src/jgikbase/idmapping/core/user_handler.py

```diff
@@ -44,12 +48,96 @@ def is_valid_user(self, username: Username) -> Tuple[bool, Optional[int], Option
         :param username: the username to check.
         :returns: a tuple of 1) a boolean describing whether the user exists or not, 2)
             a unix epoch timestamp in seconds providing an absolute limit for the cache lifetime
-            of this result, and 3) a relative cache expiration time in seconds.
-            One of 2) and 3) may be None, but not both.
+            of this result, and 3) a relative cache expiration time in seconds. If both 3) and 4)
+            are None, the process implementing the cache must make its own decisions regarding
+            the cache lifetime.
         '''
         raise NotImplementedError()
 
 
+class UserHandlerSet:
+    """
+    A container for a number of user handlers that provides caching for said handlers.
+    """
+
+    def __init__(
+            self,
+            user_handlers: Set[UserHandler],
+            cache_timer: Callable[[], int]=None,
+            cache_max_size: int=10000,
+            cache_user_expiration: int=300,
+            cache_is_valid_expiration: int=3600
+            ) -> None:
+        """
+        Create the handler set.
+
+        The cache_* parameters are mainly provided for testing purposes.
+
+        :param user_handlers: the set of user handlers to query when looking up user names from
+            tokens or checking that a provided user name is valid.
+        :param cache_timer: the timer used for cache expiration. Defaults to time.time.
+        :param cache_max_size: the maximum size of the token -> user and username -> validity
+            caches.
+        :param cache_user_expiration: the default expiration time for the token -> user cache in
+            seconds. This time can be overridden by a user handler on a per token basis.
+        :param cache_is_valid_expiration: the default expiration time for the  username ->
+            validity cache. This time can be overridden by a user handler on a per user basis.
+        """
+        no_Nones_in_iterable(user_handlers, 'user_handlers')
+        self._handlers = {handler.get_authsource_id(): handler for handler in user_handlers}
```

Comment:
maybe there is a better word than "handler" that is more descriptive? That word doesn't convey meaning to me -- you could say all programs handle something.

Reply:
Changed handler to lookup all over the code base, so e.g. LocalUserHandler is now LocalUserLookup.